Libraries/oneMKL/fourier_correlation/fcorr_2d_usm.cpp

```diff
@@ -6,151 +6,291 @@
-    oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
-                                 oneapi::mkl::dft::domain::REAL>
-        backward_plan({n_rows, n_cols});
+  // Initialize array for calculating corr without Discrete Fourier Transforms
+  // (for comparison purposes). The same column stride as for DFT-based
+  // calculations is used to satisfy the requirements of naive_cross_correlation:
+  auto naive_corr = sycl::malloc_shared<float>(n_rows * (n_cols / 2 + 1) * 2, Q);
 
-    // Data layout in conjugate-even domain
-    backward_plan.set_value(oneapi::mkl::dft::config_param::INPUT_STRIDES,
-                            complex_layout);
+  // Set the relevant image data as shifted versions of one another
+  auto evt = Q.parallel_for(sycl::range<2>{n_rows, n_cols},
+                           [=](sycl::id<2> idx) {
+    size_t row = idx[0];
+    size_t col = idx[1];
+    img1[row * col_stride_fwd_domain + col] = 0.0f;
+    img2[row * col_stride_fwd_domain + col] = 0.0f;
+  });
+  Q.single_task<>({evt}, [=]() {
+    // Set a box of unit elements in multi-indices (4-5, 5-6) for the first
+    // image
+    img1[4 * col_stride_fwd_domain + 5] = 1.0f;
+    img1[4 * col_stride_fwd_domain + 6] = 1.0f;
+    img1[5 * col_stride_fwd_domain + 5] = 1.0f;
+    img1[5 * col_stride_fwd_domain + 6] = 1.0f;
+    // Set a box of unit elements in multi-indices (1-2, 1-2) for the second
+    // image
+    img2[1 * col_stride_fwd_domain + 1] = 1.0f;
+    img2[1 * col_stride_fwd_domain + 2] = 1.0f;
+    img2[2 * col_stride_fwd_domain + 1] = 1.0f;
+    img2[2 * col_stride_fwd_domain + 2] = 1.0f;
+  }).wait();
 
-    // Data layout in real domain
-    backward_plan.set_value(oneapi::mkl::dft::config_param::OUTPUT_STRIDES,
-                            real_layout);
+  print_image(img1, Q, n_rows, n_cols, col_stride_fwd_domain, "First image:");
+  print_image(img2, Q, n_rows, n_cols, col_stride_fwd_domain, "Second image:");
+  // Calculate Frobenius norms of both input images before proceeding (for
+  // normalization purposes and for the definition of error tolerance)
+  const float norm_img1 =
+          calc_frobenius_norm(img1, Q, n_rows, n_cols, col_stride_fwd_domain);
+  const float norm_img2 =
+          calc_frobenius_norm(img2, Q, n_rows, n_cols, col_stride_fwd_domain);
+  // 1) Calculate the cross-correlation naively (for verification purposes);
+  naive_cross_correlation(Q, n_rows, n_cols, col_stride_fwd_domain,
+                          img1, img2, naive_corr).wait();
+  // 2) Calculate the cross-correlation via Discrete Fourier Transforms (DFTs):
+  //         corr = (1/num_elem) * iDFT(DFT(sig1) * CONJ(DFT(sig2)))
+  // Initialize DFT descriptor
+  oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
+                               oneapi::mkl::dft::domain::REAL>
+    desc({n_rows, n_cols});
+  desc.set_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE,
+                 1.0f / num_elem);
```

Comment:
Should we add a note about default value for forward scale?

Reply:
Done in e776e50b2647a4e7d10808cd8a9c2617005766db